### ISSUE TITLE
AIOOBE generating report from another report caused by chat attachments parents not exported

### ISSUE THREAD

**Issue by aberenguel** [sepinf-inc/IPED#1028]
IPED version: 3.18.14
OS: Ubuntu 21.10

I was generating a report from report previously generated.
This exception occured
```
2022-03-21 16:44:17	[INFO]	[indexer.desktop.ReportDialog$2]			2022-03-21 16:44:17	[ERROR]	[gpinf.indexer.IndexFiles]			Processing Error: 
java.lang.Exception: Error decoding datasource /tmp/report13429178478623615301.iped
	at dpf.sp.gpinf.indexer.datasource.ItemProducer.run(ItemProducer.java:154) ~[iped-engine-3.18.14.jar:?]
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 376749 out of bounds for length 376748
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.insertParentTreeNodes(IPEDReader.java:252) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.processBookmark(IPEDReader.java:213) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.read(IPEDReader.java:136) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.ItemProducer.run(ItemProducer.java:124) ~[iped-engine-3.18.14.jar:?]
2022-03-21 16:44:17	[INFO]	[indexer.desktop.ReportDialog$2]			ERROR!!!
```

**Comment by aberenguel** [comment]
I saved the file `/tmp/report13429178478623615301.iped` if it would be needed.

**Comment by lfcnassif** [comment]
Thanks @aberenguel for reporting. Could you do some tests to help me understand what is happening? Just increasing the array size should not fix the root cause of the issue. I pushed a code for debugging to #1028_debug branch (snapshot soon here: https://github.com/sepinf-inc/IPED/actions/runs/2018553400). Just replace the indexador/lib/iped-search-app.jar and do these 3 searches:
```
id:376749
parentId:376749
parentIds:376749
```

What are the results returned for each of them?

**Comment by lfcnassif** [comment]
PS: commit above just makes tree nodes (shown in Evidences tab) not exported to report to be listed in table tab. Actually they are indexed without contents with a special flag treeNode:true to be rendered in file tree, but not in table tab results.

**Comment by aberenguel** [comment]
The two first search returned empty. The last one returned 1 item:
![image](https://user-images.githubusercontent.com/2973359/159373860-0fb9e84a-00ca-4d18-a90b-1f5e8d60edde.png)

**Comment by aberenguel** [comment]
Got those lines in log:
```
2022-03-21 19:26:15	[INFO]	[indexer.desktop.PesquisarIndice]			Searching for query id:[376749 TO 376749]
2022-03-21 19:26:18	[INFO]	[indexer.desktop.PesquisarIndice]			Searching for query parentId:[376749 TO 376749]
2022-03-21 19:26:22	[INFO]	[indexer.desktop.PesquisarIndice]			Searching for query parentIds:376749
```

**Comment by aberenguel** [comment]
I've just replaced the iped-search-app.jar and tried to generate the report again.
Same error:
```
java.lang.Exception: Error decoding datasource /tmp/report17378613095477859714.iped
	at dpf.sp.gpinf.indexer.datasource.ItemProducer.run(ItemProducer.java:154) ~[iped-engine-3.18.14.jar:?]
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 376749 out of bounds for length 376748
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.insertParentTreeNodes(IPEDReader.java:252) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.processBookmark(IPEDReader.java:213) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.read(IPEDReader.java:136) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.ItemProducer.run(ItemProducer.java:124) ~[iped-engine-3.18.14.jar:?]
2022-03-21 19:35:24	[INFO]	[indexer.desktop.ReportDialog$2]			ERROR!!!
```

**Comment by lfcnassif** [comment]
> The first two searches returned empty. The last one returned 1 item:

After replacing the jar, right?

Are you able to find that file's parent in file tree?



> I've just replaced the iped-search-app.jar and tried to generate the report again.

I still didn't tried to fix it, just changed the code to help testing.

**Comment by lfcnassif** [comment]
Was the original case (before creating the first report) processed using blind or a similar profile (with automatic file extraction on)?

**Comment by lfcnassif** [comment]
ok I _think_ I know what could have happened. Is that file a media attachment referenced by some chat exported to report (and the media itself not explicitly included in, but included automatically because it is referenced by the chat)? Just click on it and the the `References` tab to see if some chat reference is listed.

If the answer is yes to above and no to below question

> Are you able to find that file's parent in file tree?

I know how to fix it, but you will have to generate the first report again, since the file tree should be broken, the parent of that file wasn't exported in the first generated report.

**Comment by lfcnassif** [comment]
Just pushed a possible fix.

**Comment by aberenguel** [comment]
> > The first two searches returned empty. The last one returned 1 item:
> 
> After replacing the jar, right?
Yes. Before replace the too.

> 
> Are you able to find that file's parent in file tree?
> 
The item has these metadata: `id: 376747` and `parentId: 376748`. But the parent was not there. I searched for `id:376748`, but no results.

**Comment by aberenguel** [comment]
> ok I _think_ I know what could have happened. Is that file a media attachment referenced by some chat exported to report (and the media itself not explicitly included in, but included automatically because it is referenced by the chat)? Just click on it and check the `References` tab to see if some chat reference is listed.

I think that's exactly the case. The item was exported because it was attached to a chat. It didn't have any bookmark after reporting generation.
Looking at `References` tab, the corresponding chat is there.

**Comment by aberenguel** [comment]
In the original case

![image](https://user-images.githubusercontent.com/2973359/159523356-f4539a9a-8bf7-4b4b-8d6f-a156931c9e88.png)

The generated report

![image](https://user-images.githubusercontent.com/2973359/159523416-e1b8a2ae-5310-4f70-9d66-4449ff8998f6.png)

**Comment by lfcnassif** [comment]
Ok thanks @aberenguel for checking my guesses. I think the pushed changes will fix this. Unfortunately you'll have to generate the first report again, because parents weren't exported. I tried to reproduce the issue locally, but couldn't, since this depends on ids assigning order.

About your last trace, that doesn't make sense with 3.18.4 or with last commit I pushed here. What's the exact version you are using?

**Comment by aberenguel** [comment]
> Well, I will merge the proposed fix to master. If your last error happens again, we can open a separate issue, that seems unrelated, thanks again.

Great. Just to checked in my case:
```
$ sha256sum indexador/lib/iped-engine-3.18.14.jar 
b7b3f85aa2ba55840f77bb3d7961b7aaf2343de33017f60fa03fdbde59159105  indexador/lib/iped-engine-3.18.14.jar
```

**Comment by aberenguel** [comment]
I've just generated the report using `iped-engine-3.18.14.jar` from  [iped-snapshot-64360ac2f382d500ae447b4aafa70fb9d448fcc2.zip](https://github.com/sepinf-inc/IPED/actions/runs/2022424801).
The ids assigning order has changed (as expected).

Then I proceeded the generation of another reported from that fresh report.
Got this exception:
```
2022-03-22 16:53:45	[INFO]	[indexer.desktop.ReportDialog$2]			2022-03-22 16:53:45	[ERROR]	[gpinf.indexer.IndexFiles]			Processing Error: 
java.lang.Exception: Error decoding datasource /tmp/report6438687628344223275.iped
	at dpf.sp.gpinf.indexer.datasource.ItemProducer.run(ItemProducer.java:154) ~[iped-engine-3.18.14.jar:?]
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 741484 out of bounds for length 741483
	at dpf.sp.gpinf.indexer.search.Marcadores.addLabel(Marcadores.java:169) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.copyBookmarksToReport(IPEDReader.java:244) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.processBookmark(IPEDReader.java:218) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.read(IPEDReader.java:139) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.ItemProducer.run(ItemProducer.java:124) ~[iped-engine-3.18.14.jar:?]
```

**Comment by lfcnassif** [comment]
> Got this exception:

Not sure what is happening again, seems there is an item (or parent ref) in second report with an ID greater than the max ID of first report... Maybe there is still some parent missing with a large ID being referenced, or maybe items are being expanded in report generation, they shouldn't... Your report seems large, maybe items tagged to be included in report are getting a large (unexpected) ID because new items being expanded wrongly are consuming some IDs...

I pushed some debugging code to try to debug this, if you could run the second report again with last snapshot of #1028_debug branch. If the expansion theory is right, the pushed changes may fix this...

**Comment by aberenguel** [comment]
Just tested with 115e3d5 commit.
```
java.lang.Exception: Error decoding datasource /tmp/report11837775297608258467.iped
	at dpf.sp.gpinf.indexer.datasource.ItemProducer.run(ItemProducer.java:154) ~[iped-engine-3.18.14.jar:?]
Caused by: java.lang.RuntimeException: newID=741484 oldID=677440
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.copyBookmarksToReport(IPEDReader.java:268) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.processBookmark(IPEDReader.java:222) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.IPEDReader.read(IPEDReader.java:143) ~[iped-engine-3.18.14.jar:?]
	at dpf.sp.gpinf.indexer.datasource.ItemProducer.run(ItemProducer.java:124) ~[iped-engine-3.18.14.jar:?]
```

**Comment by aberenguel** [comment]
The parentId of `id:677440` is 677441. 
![image](https://user-images.githubusercontent.com/2973359/159623463-2ed78ea3-e3e9-4569-88ee-e679137a669b.png)

It seems the parent is not present in the case.

![image](https://user-images.githubusercontent.com/2973359/159623525-0f360cd5-150c-4058-91e5-db89f417175e.png)

**Comment by aberenguel** [comment]
> It seems the parent is not present in the case.

That's OK. The parent was a folder that was not marked to be added in the generated report.

**Comment by lfcnassif** [comment]
> It seems the parent is not present in the case.

Thanks @aberenguel. But is the parent present in file tree?

Are you appending (--append) this second report to another existing report?

**Comment by lfcnassif** [comment]
I think I found the problem, commit 381494f has an issue where the item counter thread is consuming items IDs, it shouldn't. Please give the last snapshot a try and see if it fix the the second report creation.

**Comment by aberenguel** [comment]
Thanks @lfcnassif ! It seems the AIOOBE is fixed.
I've just genereated a second report without any issues using this build https://github.com/sepinf-inc/IPED/actions/runs/2028250135

A point I'd like to mention is that the second report came with 2 items in addition:
![image](https://user-images.githubusercontent.com/2973359/159765079-2836fc79-b9c7-43fa-a016-fa51eb0cf8fd.png)

**Comment by aberenguel** [comment]
> A point I'd like to mention is that the second report came with 2 items in addition: ![image](https://user-images.githubusercontent.com/2973359/159765079-2836fc79-b9c7-43fa-a016-fa51eb0cf8fd.png)

I will open another issue about it

**Comment by lfcnassif** [comment]
Thank you @aberenguel for testing!

**Comment by lfcnassif** [comment]
Closed on master by commit [30832ae](https://github.com/sepinf-inc/IPED/commit/30832ae0dccc7e65fb1a67fc6bb78a981ce67e7f)
